zanata-war/src/main/java/org/zanata/repository/LocaleRepository.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2016, Red Hat, Inc. and individual contributors as indicated by the
+ * @author tags. See the copyright.txt file in the distribution for a full
+ * listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This software is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this software; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA, or see the FSF
+ * site: http://www.fsf.org.
+ */
+package org.zanata.repository;
+
+import org.apache.deltaspike.data.api.AbstractEntityRepository;
+import org.apache.deltaspike.data.api.Repository;
+import org.zanata.common.LocaleId;
+import org.zanata.model.HLocale;
+
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Root;
+import java.io.Serializable;
+import java.util.List;
+
+/**
+ * Repository for the HLocale entity.
+ *
+ * @author Carlos Munoz <a href="mailto:camunoz@redhat.com">camunoz@redhat.com</a>
+ */
+@Repository
+public abstract class LocaleRepository
+        extends AbstractEntityRepository<HLocale, Long>
+        implements Serializable {
+
+    public abstract List<HLocale> findByActiveEqual(boolean active);
+
+    public abstract HLocale findOptionalByLocaleId(LocaleId localeId);
+
+    public List<HLocale> findByActive() {
+        return findByActiveEqual(true);
+    }
+
+    public HLocale findByLocaleId(LocaleId localeId) {
+        return findOptionalByLocaleId(localeId);
+    }
+
+    public List<HLocale> findAllActiveAndEnabledByDefault() {
+        // TODO To get rid of some of this boilerplate code and/or use
+        // DeltaSpike's criteria API, we might need to activate JPA metamodel
+        // generation
+        CriteriaQuery<HLocale> criteria = criteriaQuery();
+        CriteriaBuilder cb =
+                entityManager().getEntityManagerFactory()
+                        .getCriteriaBuilder();
+        Root<HLocale> root = criteria.from(entityClass());
+        criteria.where(
+                cb.and(
+                        cb.equal(root.get("active"), true),
+                        cb.equal(root.get("enabledByDefault"), true)
+                )
+        );
+
+        return entityManager().createQuery(criteria).getResultList();
+    }
+}
```

Comment:
On 11 February 2016 at 15:31, Carlos Munoz <notifications@github.com> wrote:

> In zanata-war/src/main/java/org/zanata/repository/LocaleRepository.java
> <https://github.com/zanata/zanata-server/pull/1083#discussion_r52565616>:
>
> > +        // generation
> > +        CriteriaQuery<HLocale> criteria = criteriaQuery();
> > +        CriteriaBuilder cb =
> > +                entityManager().getEntityManagerFactory()
> > +                        .getCriteriaBuilder();
> > +        Root<HLocale> root = criteria.from(entityClass());
> > +        criteria.where(
> > +                cb.and(
> > +                        cb.equal(root.get("active"), true),
> > +                        cb.equal(root.get("enabledByDefault"), true)
> > +                )
> > +        );
> > +
> > +        return entityManager().createQuery(criteria).getResultList();
> > +    }
> > +}
>
> So, the more complex method is because the generic JPA criteria API is
> really complex.
>
That sounds like another argument for Zanata to use Hibernate and not pure
JPA, where there's an option.

I think we can either encapsulate some of the boilerplate code, or use some
> kind of wrapper (like Deltaspike's one). We will need to activate JPA's
> metamodel generation which basically gives you compiled classes with all
> your tables and fields.
>

It might be time we used Querydsl too, which even gets a mention here:
https://deltaspike.apache.org/documentation/data.html#QueryDelegates


> We can also completely replace the DAO by moving this method into the
> repository. Since the repository is using entity manager, I would like to
> keep it that way, but there's nothing stopping us from continuing to use
> hibernate sessions.
>


-- 
Sean Flanigan

Principal Software Engineer
Globalisation Tools Engineering
Red Hat